### PR TITLE
465 animal observation summary

### LINKED ISSUE #465: Resource for individual animal observation statistics
As herd monitoring system manufacturer we need an option to facilitate the transfer of individual-level animal measurements. Together with some of our herd management system integration partners we came up with the following proposal for a new animal measurement endpoint. Our goal is to maintain the versatility of the animal measurements data structure, ensuring its compatibility with a diverse array of measurement types collected by various systems and sensors. It is essential that this endpoint is designed with foresight, eliminating the need for new endpoints with the introduction of novel measurement types.

Initially, we aim to utilize this animal measurements endpoint to export data concerning the daily time spent on behaviours like ruminating, eating, and inactivity. Later on we would also like to use the endpoint for exports of data such as heat activity values, health score values, automatically measured locomotion scores and other time series datasets.

# Proposal
## Animal Measurements endpoint
This endpoint returns data about animal measurements for all animals on a specific location. 

GET: /v1/locations/{location-scheme}/{location-id}/animal-measurements

#### Path
-	location-scheme: location identifier scheme
-	location-id: id

#### Query 
-	measurement-types: List of measurement types that you want to request, see measurement types Enum below
-	measurement-datetime-from: The start datetime of the range in which you want to request data (UTC)
-	measurement-datetime-to: The end datetime of the range in which you want to request data (UTC)
-	page-size: The number of items to display on one page of results. 
-	page: The page number
-	animal-scheme: Select the animal scheme
-	animal-id: Select a specific animal

### Response Schema

- **animalMeasurements** [list]
  - **animal**
    - scheme [string]: Identfies the data source as com.manufacturer.animal.id
    - id [string]: The unique Manufacturer animal identifier (UUID)
  - **measurements** [list]
    - **type** [enum]: The measurement type
      - DailyRuminationTime
      - DailyEatingTime
      - DailyInactiveTime
      - Last24HourRuminationTime
      - Last24HourEatingTime
      - Last24HourInactiveTime
    - **value**: The measurement value
    - **valueType** [enum]: the type of the measurement values
      - Number
    - **unit** [enum]: the unit of the measurement, UN/CEFACT Common Code for Units of Measurement.
      - SEC
      - MIN
    - **isComplete** [boolean] (optional): Flag that indicates if the measurement is complete. For daily measurements this flag is complete if the value is based on a full day of measurements. If not all measurements have been received the value will be false.
    - **dateTime** [datetime]: The start date and time of the measurement in UTC.

### Response example

```json
{
  "view": {
    "totalItems": 0,
    "totalPages": 0,
    "pageSize": 0,
    "currentPage": 0
  },
  "member": [
    {
      "resourceType": "string",
      "self": "string",
      "meta": {
        "source": "string",
        "sourceId": "string",
        "isDeleted": true,
        "modified": "2024-02-07T22:14:07.635Z",
        "created": "2024-02-07T22:14:07.635Z",
        "creator": "string",
        "validFrom": "2024-02-07T22:14:07.635Z",
        "validTo": "2024-02-07T22:14:07.635Z"
      },
      "location": {
        "scheme": "com.nedap.farm.id",
        "id": "2f7c3d10-4d2c-4758-81b4-4d8f496ea659"
      },
      "animalMeasurements": [
        {
          "animal": {
            "scheme": "com.nedap.animal.id",
            "id": "fd1b3d96-2620-4040-a1a2-8d913b11b788"
          },
          "measurements": [
            {
              "type": "DailyRuminationTime",
              "value": 400,
              "valueType": "Number",
              "unit": "MIN",
              "isComplete": true,
              "dateTime": "2024-02-07T23:00:00.000Z"
            },
            {
              "type": "DailyEatingTime",
              "value": 300,
              "valueType": "Number",
              "unit": "MIN",
              "isComplete": true,
              "dateTime": "2024-02-07T23:00:00.000Z"
            },
            {
              "type": "DailyInactiveTime",
              "value": 500,
              "valueType": "Number",
              "unit": "MIN",
              "isComplete": true,
              "dateTime": "2024-02-07T23:00:00.000Z"
            },
            {
              "type": "Last24HourRuminationTime",
              "value": 350,
              "valueType": "Number",
              "unit": "MIN",
              "isComplete": true,
              "dateTime": "2024-02-08T16:14:07.635Z"
            }
          ]
        }
      ]
    }
  ]
}
```

### PR DESCRIPTION
**Pull request by cookeac**
I tried to add proposed commits to #467 but didn't have permissions. 
This is a draft PR for review and comments. It addresses recommendations from the ADE meetings to address #465.

- The initial examples are (time-based) summaries of observations (e.g. the number of minutes in a day spent in an activity, or the number of steps in a day).
- The `icarStatisticsType` (used in group-based `icarStatisticsResource`) is close to what we need but we need to address time period or duration
- Address naming conventions - avoid the use of bare "type" and consider using "metric" as we use this for similar things elsewhere.
- Recognise that we don't need "DailyRuminationTime" and "24HourRuminationTime" - these are both different duration or period summaries for the same metric.

If we are on the right track, I will copy in the URL scheme that has already been defined for #467.

### CODE REVIEW

**Comment by erwinspeybroeck** [comment]
1) typo in the name of resources/icar**l**ObservationSummaryResource.json

2) a lot of redundancy?

for each animal the location/dateFrom/dateTo --> is it possible that this changes within the same response?

for each statistic the startDateTime and duration --> is it possible that this changes within the same response?

I think the response will always contain data for the same location, dates, periods, so 

{ location, date from, dateTo, startDateTime, duration (or you use dateFrom - dateTo, or you use startDateTime + duration)
within this an array of animals
within each animal an array of statistics

we also add an id on the main level - don't know if that is necessary

for the animal we alsi had a request to add the parity at the moment of the observation - i'm not 100% in favor, but ...

Of course a receiver will probably store this  in a "flat" way on animal level:
animal, location, date info, metric info --> so without any arrays. Don't know if we have to take that into account.

**Comment by cookeac** [comment]
> 1. typo in the name of resources/icar**l**ObservationSummaryResource.json

Oops, will fix

> 2. a lot of redundancy?

Definitely there is redundancy in the `icarObservationSummaryResource` having location, dateFrom, and dateTo which will also be parameters to the API call. These could be removed entirely to my mind, but I am keen for other people's thoughts.

The statistics are an array of "daily" data (or whatever periods are provided), so startDateTime of a statistic will vary for every statistic for an animal. Similarly the duration may vary for each statistic, returning (for instance) "D" or the "H24" in the same API call. So there is no real redundancy for these.

File: enums/icarDurationType.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+  "type": "string",
+
+  "enum": [
+    "1D",
+    "1H",
+    "24H",
+    "1W",
+    "1M"
+  ]
```
Comment (cookeac) [inline enums/icarDurationType.json:13]: Add 96H for compatibility with ICAR recommendations

File: resources/icarObservationSummaryResource.json [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        "type": "object",
+  
+        "required": ["animal"],
+  
+        "properties": {
+          "animal": {
+            "$ref": "../types/icarAnimalIdentifierType.json",
+            "description": "Unique animal scheme and identifier combination."
+          },
+          "location": {
```
Comment (cookeac) [inline resources/icarObservationSummaryResource.json]: Remove location - as this is in icarResource

File: resources/icarObservationSummaryResource.json [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+            "$ref": "../types/icarLocationIdentifierType.json",
+            "description": "Location where the observations occurred."
+          },
+          "dateFrom": {
```
Comment (cookeac) [inline resources/icarObservationSummaryResource.json]: remove dateFrom and dateTo as they are in the API query parameters

File: collections/icarObservationSummaryCollection.json [hunk 4]
```diff
@@ -0,0 +1,21 @@
+{
+  "description": "Represents a collection of animal sets. Based on icarResourceCollection to provide paging etc.",
```
Comment (NedapLivestockManagement) [inline collections/icarObservationSummaryCollection.json]: should be: Represents a collection of observation summary statistics.

**Review by erwinspeybroeck — APPROVED** [review]

File: collections/icarObservationSummaryCollection.json (hunk 4 above)
Comment (cookeac) [inline collections/icarObservationSummaryCollection.json]:
```suggestion
  "description": "Represents a collection of observation summary statistics. Based on icarResourceCollection to provide paging etc.",
```

**Comment by cookeac** [comment]
Thanks for the feedback and adjustments. Based on our discussion on Nov 27, 2004 we are merging this.
Resolves #465

**Pull request merged**